pkg/api/aim/encoding/decoder.go

```diff
@@ -32,17 +32,174 @@ func (d *reader) readField() ([]byte, error) {
 	bufferLength := make([]byte, 4)
 	_, err := io.ReadFull(d, bufferLength)
 	if err != nil {
-		return nil, eris.Wrap(err, "error reading data into the buffer")
+		return nil, eris.Wrap(err, "Error reading Data into the buffer")
 	}
 	data := make([]byte, binary.LittleEndian.Uint32(bufferLength))
 	_, err = io.ReadFull(d, data)
 	if err != nil {
-		return nil, eris.Wrap(err, "error reading data into the buffer")
+		return nil, eris.Wrap(err, "Error reading Data into the buffer")
 	}
 	return data, nil
 }
 
-// Decode decodes input stream of data into map[string]interface{}.
+// DecoderResult represents result of one piece of data after decoding.
+type DecoderResult struct {
+	Data    map[string]interface{}
+	Error   error
+	Portion int
+}
+
+// DecoderProvider provides an interface to work with Decoder.
+type DecoderProvider interface {
+	// Decode represents syntactic sugar function which returns decoded stream data at once.
+	// this function is just for back compatibility with integration tests.
+	Decode() (map[string]interface{}, error)
+	// DecodeByChunk decodes stream data by chunks.
+	DecodeByChunk() <-chan DecoderResult
```

Comment:
I _think_ that we should have an interface that is closer to an iterator (like in the GCS and S3 APIs). Exposing a raw channel is a bit too low-level IMO. The underlying implementation can totally use a channel if you'd like, though I am not convinced of the actual performance benefits of it. The implementation may actually be simpler and more readable without channels.

Reply:
ok, let me changes it then.

Reply:
ok, I think now it should be similar with S3 or GCS.